### PR TITLE
fix: send flight keep-alives

### PR DESCRIPTION
**Pull request by crepererum**
This was tested using a local nginx proxy and a small modication to the IOx source code: insert `make_stream_slow` into `GetStream::new` just between `let query_results = ...` and
`let innr = FlightDataEncoderBuilder::new()...`.

Fixes https://github.com/influxdata/idpe/issues/17824 .

### CODE REVIEW

File: service_grpc_flight/src/keep_alive.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+//!     Received RST_STREAM with error code 2. gRPC client debug context:
+//!       UNKNOWN:Error received from peer ipv6:%5B::1%5D:8888 {
+//!         created_time:"2023-07-03T17:54:56.346363565+02:00",
+//!         grpc_status:13,
+//!         grpc_message:"Received RST_STREAM with error code 2"
+//!       }.
+//!       Client context: OK
+//! ```
+//!
+//! `Received RST_STREAM with error code 2` is a good hint. A to [RFC 7540] (the HTTP/2 spec) the error code is (see
```
Comment (mhilton) [inline service_grpc_flight/src/keep_alive.rs:36]: I'm not sure what you mean by "A to [RFC 7540]..."

Comment (crepererum) [reply service_grpc_flight/src/keep_alive.rs:36]: a -> according (typo)

File: service_grpc_flight/src/keep_alive.rs [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+    use futures::TryStreamExt;
+    use test_helpers::maybe_start_logging;
+
+    use super::{test_util::make_stream_slow, *};
+
+    type BatchStream = BoxStream<'static, Result<RecordBatch, FlightError>>;
+    type FlightStream = BoxStream<'static, Result<FlightData, FlightError>>;
+
+    #[tokio::test]
+    #[should_panic(expected = "stream timeout")]
```
Comment (NGA-TRAN) [inline service_grpc_flight/src/keep_alive.rs:312]: Nice

File: service_grpc_flight/src/keep_alive.rs [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+    #[tokio::test]
+    async fn test_keep_alive() {
+        maybe_start_logging();
+
+        let s = make_test_stream(true);
+        let s = FlightRecordBatchStream::new_from_flight_data(s);
+        let batches: Vec<_> = s.try_collect().await.unwrap();
+        assert_batches_eq!(
+            vec!["+---+", "| f |", "+---+", "| 1 |", "| 2 |", "| 3 |", "| 4 |", "| 5 |", "+---+",],
+            &batches
```
Comment (NGA-TRAN) [inline service_grpc_flight/src/keep_alive.rs:328]: 🎉

**Review by NGA-TRAN — APPROVED** [review]

File: service_grpc_flight/src/keep_alive.rs [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+impl KeepAliveStream {
+    /// Create new keep-alive wrapper from the underlying stream and the given interval.
+    ///
+    /// The interval is measured from the last message -- which can either be a "real" message or a keep-alive.
+    pub fn new<S>(s: S, interval: Duration) -> Self
+    where
+        S: Stream<Item = Result<FlightData, FlightError>> + Send + 'static,
+    {
+        let mut ticker = tokio::time::interval(interval);
+        ticker.set_missed_tick_behavior(MissedTickBehavior::Delay);
```
Comment (alamb) [inline service_grpc_flight/src/keep_alive.rs:158]: https://docs.rs/tokio/1.29.1/tokio/time/enum.MissedTickBehavior.html#variant.Delay -- seems reasonable

**Review by alamb — APPROVED** [review]
👏  Very nice

**Pull request merged**